### PR TITLE
feat: added sorting and delta sync

### LINKED ISSUE #19: Feature: Handle Microsoft Graph Delta Sync with Ordered Changes
### 📝 Description
When receiving webhook notifications from Microsoft Graph (e.g., for calendar events), we use the delta API (`/events/delta`) to fetch incremental changes. However, Microsoft does **not guarantee the order** of the returned changes.

To ensure deterministic processing, we need to **sort the changes manually** based on their `lastModifiedDateTime`.

---

### ✅ Tasks

- [ ] Retrieve `lastModifiedDateTime` (or fallback to `createdDateTime`) from delta API response.
- [ ] Sort the list of returned events/changes by that timestamp.
- [ ] Ensure the delta sync process stores and uses per-user delta links correctly. (Sync with DB)
- [ ] Test across multiple users and edge cases (e.g., simultaneous updates).

---

### 📌 References

- [MS Graph Delta API - docs](https://learn.microsoft.com/en-us/graph/delta-query-overview)
- [Webhooks + Delta Query Pattern](https://learn.microsoft.com/en-us/graph/webhooks-with-resource-data)

---

### 🧠 Notes

- Order is not guaranteed by the API response. Manual sorting is essential for consistency.
- Store delta links per user to avoid missing changes.

### PR DESCRIPTION
**Pull request by Gmin2**
https://github.com/user-attachments/assets/187ceec3-7a1a-4a25-bb49-fc0427cc531d

Fixes #19 
/claim #19

### CODE REVIEW

File: src/services/calendar/calendar.service.ts [hunk 1]
```diff
@@ -440,4 +455,92 @@ export class CalendarService {
...
+
+      // If we have a delta link, use it
+      if (deltaLink) {
+        requestUrl = deltaLink;
+      }
+
+      const allEvents: Event[] = [];
+
+      // Fetch all pages of changes
+      let response: any = { '@odata.nextLink': requestUrl };
```
Comment (ramielsawy) [inline src/services/calendar/calendar.service.ts:487]: Let's type this by creating an interface or a type for it

File: src/services/calendar/calendar.service.ts [hunk 2]
```diff
@@ -440,4 +455,92 @@ export class CalendarService {
...
   }
-}
\ No newline at end of file
+
+  /**
+   * Fetch changes from the Delta API and sort them by lastModifiedDateTime
+   * @param externalUserId - External user ID
+   * @returns Array of events sorted by lastModifiedDateTime
+   */
+  async fetchAndSortChanges(externalUserId: string): Promise<Event[]> {
```
Comment (ramielsawy) [inline src/services/calendar/calendar.service.ts:464]: Let's move this to a shared service (e.g. `delta-sync.service.ts` in a `shared` folder next email and calendar folders) and have it handle both `messages` and `events`

You could have something like 

```ts
export interface DeltaItem {
  lastModifiedDateTime?: string;
  createdDateTime: string;
}

export type DeltaEvent = Event & DeltaItem;
export type DeltaMessage = Message & DeltaItem;
```

and have something like

```ts
  async fetchAndSortChanges<T extends DeltaItem>(
```

File: src/services/calendar/calendar.service.ts [hunk 3]
```diff
@@ -440,4 +455,92 @@ export class CalendarService {
...
+      // Fetch all pages of changes
+      let response: any = { '@odata.nextLink': requestUrl };
+
+      while (response['@odata.nextLink']) {
+        const nextLink = response['@odata.nextLink'];
+
+        if (nextLink.startsWith('https://')) {
+          response = await client.api(nextLink).get();
+        } else {
+          response = await client.api(nextLink).get();
```
Comment (ramielsawy) [inline src/services/calendar/calendar.service.ts:495]: Looks unnecessary

File: src/services/calendar/calendar.service.ts [hunk 4]
```diff
@@ -440,4 +455,92 @@ export class CalendarService {
...
+
+      return sortedEvents;
+    } catch (error: unknown) {
+      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
+      this.logger.error(`Failed to fetch calendar changes: ${errorMessage}`);
+      throw new Error(`Failed to fetch calendar changes: ${errorMessage}`);
+    }
+  }
+
+  async getExternalUserIdFromUserId(userId: number): Promise<string | null> {
```
Comment (ramielsawy) [inline src/services/calendar/calendar.service.ts:530]: IF you need it (and I don't think you'll need it for retrieving the delta events), then maybe call the repository directly instead of wrapping the function with no extra logic

File: src/services/email/email.service.ts
Comment (ramielsawy) [inline src/services/email/email.service.ts:1]: email service is missing the implementation of delta links

**Review by ramielsawy — CHANGES_REQUESTED** [review]
Thank you so much @Gmin2 for your contribution. I left a few comments. One open question other than the comments would be how do we handle delays, replays, sync reset and token expiration. Refer to those sections [here](https://learn.microsoft.com/en-us/graph/delta-query-overview#processing-delays)

If you ever want to discuss this here's my scheduling link https://app.reclaim.ai/m/checkfirst/rami

Thank you

**Comment by ramielsawy** [comment]
> can you review it once @ramielsawy ?

Hi @Gmin2 , I updated your branch to be up to date with main (requirement for merging, please don't force push on it). 

Just two notes:

1. There's a [build failure](https://github.com/checkfirst-ltd/nestjs-outlook/actions/runs/15874252887/job/44757980586?pr=23)
2. Not sure all the comments I left were taken care of, can you please go through them and comment on them with "done"

Thanks

File: src/services/calendar/calendar.service.ts (hunk 1 above)
Comment (Gmin2) [reply src/services/calendar/calendar.service.ts]: done in https://github.com/checkfirst-ltd/nestjs-outlook/pull/23/commits/b391cefe2a04a0e9d1f6c29ac6c5cf508c1c8bd5

File: src/services/calendar/calendar.service.ts (hunk 2 above)
Comment (Gmin2) [reply src/services/calendar/calendar.service.ts:467]: done in https://github.com/checkfirst-ltd/nestjs-outlook/pull/23/commits/b391cefe2a04a0e9d1f6c29ac6c5cf508c1c8bd5

File: src/services/email/email.service.ts
Comment (Gmin2) [reply src/services/email/email.service.ts:1]: implemented

**Review by ramielsawy — CHANGES_REQUESTED** [review]
Thank you so much @Gmin2 we finally got to the end of it

1. Here's a lint failure https://github.com/checkfirst-ltd/nestjs-outlook/actions/runs/15880265218/job/44778720262?pr=23
2. `async getExternalUserIdFromUserId` is repeated twice when it's not related to any of both services. Maybe have it done once in a shared service or something.

**Comment by Gmin2** [comment]
cc @ramielsawy

File: src/services/email/email.service.ts
Comment (ramielsawy) [reply src/services/email/email.service.ts:1]: Hi @Gmin2 , I am not sure the [email service](https://github.com/checkfirst-ltd/nestjs-outlook/pull/23/files#diff-9c4893783741aaec5572e951f222f9d5653126ca3d6fec84814416fb2e78ff7d) has any new implementation.
